flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/RecoverableAtomicFileWriter.java

```diff
@@ -0,0 +1,97 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.file.utils;
+
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.core.fs.RecoverableFsDataOutputStream;
+import org.apache.flink.core.fs.RecoverableWriter;
+
+import java.io.IOException;
+
+/** A {@link AtomicFileWriter} to wrap {@link RecoverableWriter}. */
+public class RecoverableAtomicFileWriter implements AtomicFileWriter {
+
+    private final RecoverableWriter recoverableWriter;
+
+    public RecoverableAtomicFileWriter(RecoverableWriter recoverableWriter) {
+        this.recoverableWriter = recoverableWriter;
+    }
+
+    @Override
+    public AtomicFsDataOutputStream open(Path path) throws IOException {
+        RecoverableFsDataOutputStream out = recoverableWriter.open(path);
+        return new RecoverableAtomicFsDataOutputStream(out);
+    }
+
+    private static class RecoverableAtomicFsDataOutputStream extends AtomicFsDataOutputStream {
+
+        private final RecoverableFsDataOutputStream out;
+
+        public RecoverableAtomicFsDataOutputStream(RecoverableFsDataOutputStream out) {
+            this.out = out;
+        }
+
+        @Override
+        public void write(int b) throws IOException {
+            this.out.write(b);
+        }
+
+        @Override
+        public void write(byte[] b, int off, int len) throws IOException {
+            this.out.write(b, off, len);
+        }
+
+        @Override
+        public void flush() throws IOException {
+            this.out.flush();
+        }
+
+        @Override
+        public void sync() throws IOException {
+            this.out.sync();
+        }
+
+        @Override
+        public long getPos() throws IOException {
+            return this.out.getPos();
+        }
+
+        @Override
+        public boolean closeAndCommit() throws IOException {
+            RecoverableFsDataOutputStream.Committer committer = this.out.closeForCommit();
+            committer.commit();
+
+            /*
+             * When using RecoverableWriter, we can only return true, and there are
+             * two cases where commit actually fails.
+             * 1. Commit has an exception, here it will not return true, throwing an
+             *    exception for the outer caller to handle.
+             * 2. Commit fails but there is no exception. This case requires the outer
+             *    caller to have a lock to ensure that there is no concurrent writing of
+             *    the current file.
```

Comment:
This case happens for object stores. Why not read the file after commit and check if it is the actual file we've written?